From f638889c5825a2b993a5ba990a612716ff4fb372 Mon Sep 17 00:00:00 2001
From: mayi <3363523343@qq.com>
Date: Wed, 21 Oct 2026 18:34:15 +0800
Subject: [PATCH] fix(osdrv/atbm6461): stop firmware setup on register access
 errors

Check AHB/control accesses and return the original transport error. Do not call irq_unsubscribe before an IRQ subscription exists.
---
 .../realtek/atbm6461/hal_apollo/hwio_sdio.c   | 31 +++++++++----------
 1 file changed, 15 insertions(+), 16 deletions(-)

diff --git a/extdrv/wireless/realtek/atbm6461/hal_apollo/hwio_sdio.c b/extdrv/wireless/realtek/atbm6461/hal_apollo/hwio_sdio.c
index 62213fea9d46e4e55ae17cc43c56d72e4e3528ce..360ff19917ae0f09ef3c26b7ff3f4a63913a50f6 100644
--- a/extdrv/wireless/realtek/atbm6461/hal_apollo/hwio_sdio.c
+++ b/extdrv/wireless/realtek/atbm6461/hal_apollo/hwio_sdio.c
@@ -1299,14 +1299,18 @@ int atbm_after_load_firmware(struct atbm_common *hw_priv)
 
 	//enable gpio irq register,may need move to lmac/apb.c	SMU_Init
 	ret=atbm_ahb_read_32(hw_priv,0x161000ac,&val32);
+	if (ret < 0) {
+		atbm_printk_err("%s: AHB read 0x161000ac failed: %d\n",
+			       __func__, ret);
+		goto out;
+	}
 	val32&=0xFFFFF7F8;
 	val32|=BIT(12);
 	ret=atbm_ahb_write_32(hw_priv,0x161000ac,val32);
 	if(ret<0){
-		atbm_dbg(ATBM_APOLLO_DBG_ERROR,
-			"%s: enable_irq: can't read " \
-			"config register.\n", __func__);
-
+		atbm_printk_err("%s: AHB write 0x161000ac failed: %d\n",
+			       __func__, ret);
+		goto out;
 	}
 
 #if (PROJ_TYPE==ARES_B)
@@ -1345,7 +1349,7 @@ int atbm_after_load_firmware(struct atbm_common *hw_priv)
 			atbm_dbg(ATBM_APOLLO_DBG_ERROR,
 				"%s: enable_irq: can't read " \
 				"config register.\n", __func__);
-			goto unsubscribe;
+			goto out;
 		}
 		val32 |= ATBM_HIFREG_CONF_IRQ_RDY_ENABLE;
 		val32 &= ~ATBM_HIFREG_CONFIG_CPU_RESET_BIT;
@@ -1356,7 +1360,7 @@ int atbm_after_load_firmware(struct atbm_common *hw_priv)
 			atbm_dbg(ATBM_APOLLO_DBG_ERROR,
 				"%s: enable_irq: can't write " \
 				"config register.\n", __func__);
-			goto unsubscribe;
+			goto out;
 		}
 #if (PROJ_TYPE==ARES_B)
 		ret=atbm_ahb_write_32(hw_priv,0x16100074,0x1);
@@ -1379,7 +1383,7 @@ int atbm_after_load_firmware(struct atbm_common *hw_priv)
 			atbm_dbg(ATBM_APOLLO_DBG_ERROR,
 				"%s: set_mode: can't read config register.\n",
 				__func__);
-			goto unsubscribe;
+			goto out;
 		}
 		ret = atbm_reg_write_32(hw_priv, ATBM_HIFREG_CONFIG_REG_ID,
 			val32 & ~ATBM_HIFREG_CONFIG_ACCESS_MODE_BIT);
@@ -1387,12 +1391,14 @@ int atbm_after_load_firmware(struct atbm_common *hw_priv)
 			atbm_dbg(ATBM_APOLLO_DBG_ERROR,
 				"%s: set_mode: can't write config register.\n",
 				__func__);
-			goto unsubscribe;
+			goto out;
 		}
 		/* Unless we read the CONFIG Register we are
 		 * not able to get an interrupt */
 		mdelay(10);
-		atbm_reg_read_32(hw_priv, ATBM_HIFREG_CONFIG_REG_ID, &val32);
+		ret = atbm_reg_read_32(hw_priv, ATBM_HIFREG_CONFIG_REG_ID, &val32);
+		if (ret < 0)
+			goto out;
 		
 		//hare sdio + iot,let lmac know host initial done .add by wp
 		/* Set wakeup bit in device */
@@ -1416,11 +1422,4 @@ int atbm_after_load_firmware(struct atbm_common *hw_priv)
 	out:
 		return ret;
 
-	unsubscribe:
-		hw_priv->sbus_ops->irq_unsubscribe(hw_priv->sbus_priv);
-		return ret;
-
 }
-
-
-
